File: fastapi_book_server/app/serializers/author.py
from datetime import date
from typing import Optional

from fastapi_pagination import Page
from pydantic import BaseModel

from app.serializers.sequence import Sequence


class Author(BaseModel):
    id: int

    first_name: str
    last_name: str
    middle_name: Optional[str]

    annotation_exists: bool


class AuthorBook(BaseModel):
    id: int
    title: str
    lang: str
    file_type: str
    available_types: list[str]
    uploaded: date
    translators: list[Author]
    sequences: list[Sequence]
    annotation_exists: bool


class TranslatedBook(BaseModel):
    id: int
    title: str
    lang: str
    file_type: str
    available_types: list[str]
    authors: list[Author]
    sequences: list[Sequence]
    annotation_exists: bool


class PageWithAuthorBook(Page[AuthorBook]):
    parent_item: Author | None


class PageWithTranslatedBook(Page[TranslatedBook]):
    parent_item: TranslatedBook | None

File: src/serializers/author.rs
use chrono::NaiveDate;
use serde::Serialize;

use super::date::naive_date_serializer;
use super::sequence::Sequence;

#[derive(sqlx::FromRow, sqlx::Type, Serialize)]
pub struct Author {
    pub id: i32,
    pub first_name: String,
    pub last_name: String,
    pub middle_name: String,
    pub annotation_exists: bool,
}

#[derive(sqlx::FromRow, Serialize)]
pub struct AuthorBook {
    pub id: i32,
    pub title: String,
    pub lang: String,
    pub file_type: String,
    pub year: i32,
    pub available_types: Vec<String>,
    #[serde(serialize_with = "naive_date_serializer::serialize")]
    pub uploaded: NaiveDate,
    pub translators: sqlx::types::Json<Vec<Author>>,
    pub sequences: sqlx::types::Json<Vec<Sequence>>,
    pub annotation_exists: bool,
}
